### PR TITLE
zed: add hotplug support for spare vdevs

### PR DESCRIPTION
**Pull request by ixhamza**
### Motivation and Context

Add hotplug support for spare vdevs
### Description

This commit supports for spare vdev hotplug. The spare vdev associated with all the pools will be marked as "Removed" when the drive is physically detached and will become "Available" when the drive is reattached. Currently, the spare vdev status does not change on the drive removal and the same is the case with reattachment.

### How Has This Been Tested?

By adding and removing spares through QEMU-emulated drives.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [ ] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

**Comment by ixhamza** [comment]
@amotin, @freqlabs.

**Comment by ixhamza** [comment]
Right now I am removing spares completely from the pool on the removal event, which does not seem right after a discussion with @amotin. I will update the PR to change the spare state to REMOVED when detach and will make it AVAILABLE if the disk 
 is reattached. Making this PR Draft at the moment.

**Comment by ixhamza** [comment]
I have updated the PR. Spare vdev now transitions to the "Removed" state when the drive is removed and becomes "Available" when reattached.

**Comment by freqlabs** [comment]
Would it be a problem to just remove the checks for it being a spare in libzfs instead of changing the function signatures?

**Comment by ixhamza** [comment]
@freqlabs Apparently, it should not be a problem. I will run perform some manual testing to check for corner cases. Thank you.

**Comment by ixhamza** [comment]
@freqlabs - I have checked and there is no problem in removing the spare checks in ```zpool_vdev_remove_wanted()``` and I will remove that. For the second check, ```zpool_vdev_online()```, I don't think we should change that as it seems like we don't allow doing offline/online spare vdevs from the command line, e.g., ```zpool offline pool vdev``` and ```zpool online pool vdev```.

**Comment by freqlabs** [comment]
I looks like we could add a `ZFS_ONLINE_SPARE` flag to pass through the existing `flags` param and avoid the API change.

**Comment by ixhamza** [comment]
@freqlabs Yes, you are right. I have updated the PR. Thank you.

File: cmd/zed/agents/zfs_retire.c [hunk 1]
```diff
@@ -343,11 +406,29 @@ zfs_retire_recv(fmd_hdl_t *hdl, fmd_event_t *ep, nvlist_t *nvl,
...
+			if (strcmp(devtype, VDEV_TYPE_SPARE) == 0)
+				spare = 1;
+			else if (strcmp(devtype, VDEV_TYPE_L2CACHE) == 0)
+				l2arc = 1;
+		}
+
+		if (nvlist_lookup_uint64(nvl,
+		    FM_EREPORT_PAYLOAD_ZFS_VDEV_GUID, &vdev_guid) != 0) {
+			return;
+		}
```
Comment (freqlabs) [inline cmd/zed/agents/zfs_retire.c:420]: We typically omit braces for a single line in ZFS, as seen elsewhere in this function.

Comment (ixhamza) [reply cmd/zed/agents/zfs_retire.c]: Thank you for pointing that out. I wasn't very much sure about multiline conditions, updated it.

**Review by freqlabs — APPROVED** [review]

**Review by tonyhutter — APPROVED** [review]

**Review by amotin — APPROVED** [review]
Thanks for fixing the couple things we discussed. Now looks good to me.

**Comment by ixhamza** [comment]
@behlendorf I just rebased this PR with the latest master. It would be appreciated if you can review/approve it.

**Review by behlendorf — APPROVED** [review]

**Pull request merged**